### PR TITLE
Add support for type aliases/abbreviations

### PR DESCRIPTION
**Pull request by OceanOak**
Changelog:

```
Language
- Support type aliases
```
Follow-ups
- [ ] Deal with alias of record type in SqlCompiler
- [ ] Respect type args

### CODE REVIEW

File: backend/src/LibExecution/DvalReprInternalQueryable.fs [hunk 1]
```diff
@@ -228,6 +228,7 @@ let parseJsonV0
     | TCustomType (typeName, args), JsonValueKind.Object ->
       match Map.tryFind typeName availableTypes with
       | None -> Exception.raiseInternal "Type not found" [ "typeName", typeName ]
+      | Some (CustomType.Alias (f1)) -> convert f1 j
```
Comment (OceanOak) [inline backend/src/LibExecution/DvalReprInternalQueryable.fs:231]: TODO: find where this is supposed to be tested

**Comment by OceanOak** [comment]
I have created the following function to get the underlying type from the type alias and used it in the unify function 

**Questions:**
- Am I possibly heading in the wrong direction with this approach?
- Through debugging, I've noticed that the tests reach the exhaustiveness check. What confuses me is that despite reaching this check, the tests are still passing and not throwing the error contained within the check. Could you please clarify this?

**Function**
```
let rec getUnderlyingTypeFromAlias
  (availableTypes : Map<FQTypeName.T, CustomType.T>)
  (typ : TypeReference)
  : TypeReference =
  match typ with
  | TCustomType (typeName, typeArgs) ->
    match Map.tryFind typeName availableTypes with
    | Some (CustomType.Alias (TCustomType (innerTypeName, _))) ->
      getUnderlyingTypeFromAlias availableTypes (TCustomType(innerTypeName, typeArgs))
    | _ -> typ
  | _ -> typ
```

**Usage:**
```
let rec unify
  (path : List<string>)
  (availableTypes : Map<FQTypeName.T, CustomType.T>)
  (expected : TypeReference)
  (value : Dval)
  : Result<unit, Error.T> =

  let resolvedType = getUnderlyingTypeFromAlias availableTypes expected

  match (resolvedType, value) with
  | TVariable _, _ -> Ok() 
  ...
  | TCustomType (typeName, typeArgs), value ->
    match Map.tryFind typeName availableTypes with
    | None -> Error(TypeLookupFailure(typeName, List.rev path))
    | Some ut ->
      let err =
        Error(
          TypeUnificationFailure(
            { expectedType = expected; actualValue = value },
            List.rev path
          )
        )

      match ut, value with
      | CustomType.Alias aliasType, typ ->
        debuG "innerType" (aliasType, typ)
        let resolvedAliasType = resolveTypeAlias availableTypes aliasType
        debuG "resolvedAliasType" resolvedAliasType
        unify path availableTypes resolvedAliasType typ
        ...
```
**Test**
```
  type Id = String
  type Something = { ID : Id; Data : String }
  Something { ID = 2; Data = "test" } = Something { ID = 1; Data = "test" }
  ```

**Comment by OceanOak** [comment]
I have created the following function to get the underlying type from the type alias and used it in the Typechecker.unify function 

**Questions:**
- Am I possibly heading in the wrong direction with this approach?
- Through debugging, I've noticed that the tests reach the exhaustiveness check. What confuses me is that despite reaching this check, the tests are still passing and not throwing the error contained within the check. Could you please clarify this?

**Function**
```
let rec getUnderlyingTypeFromAlias
  (availableTypes : Map<FQTypeName.T, CustomType.T>)
  (typ : TypeReference)
  : TypeReference =
  match typ with
  | TCustomType (typeName, typeArgs) ->
    match Map.tryFind typeName availableTypes with
    | Some (CustomType.Alias (TCustomType (innerTypeName, _))) ->
      getUnderlyingTypeFromAlias availableTypes (TCustomType(innerTypeName, typeArgs))
    | _ -> typ
  | _ -> typ
```

**Usage:**
```
let rec unify
  (path : List<string>)
  (availableTypes : Map<FQTypeName.T, CustomType.T>)
  (expected : TypeReference)
  (value : Dval)
  : Result<unit, Error.T> =

  let resolvedType = getUnderlyingTypeFromAlias availableTypes expected

  match (resolvedType, value) with
  | TVariable _, _ -> Ok() 
  ...
  | TCustomType (typeName, typeArgs), value ->
    match Map.tryFind typeName availableTypes with
    | None -> Error(TypeLookupFailure(typeName, List.rev path))
    | Some ut ->
      let err =
        Error(
          TypeUnificationFailure(
            { expectedType = expected; actualValue = value },
            List.rev path
          )
        )

      match ut, value with
      | CustomType.Alias aliasType, typ ->
        let resolvedAliasType = getUnderlyingTypeFromAlias availableTypes aliasType
        unify path availableTypes resolvedAliasType typ
        ...
```
**Test**
```
  type Id = String
  type Something = { ID : Id; Data : String }
  Something { ID = 2; Data = "test" } = Something { ID = 1; Data = "test" }
  ```

File: backend/src/StdLibExecution/Libs/NoModule.fs [hunk 2]
```diff
@@ -31,15 +49,20 @@ let rec equals (a : Dval) (b : Dval) : bool =
...
+    | Some t1, Some t2 ->
+      let tn1 = getUnderlyingTypeFromAlias availableTypes t1
+      let tn2 = getUnderlyingTypeFromAlias availableTypes t2
+      tn1 = tn2
+      && Map.count a = Map.count b
+      && Map.forall
+           (fun k v ->
+             Map.tryFind k b |> Option.map (equals v) |> Option.defaultValue false)
+           a
+    | _ -> Exception.raiseCode "TODO"
```
Comment (OceanOak) [inline backend/src/StdLibExecution/Libs/NoModule.fs:62]: Change to raiseInternal and add an error message

File: backend/testfiles/execution/json.tests [hunk 3]
```diff
@@ -315,6 +316,33 @@ Json.parse<Person> """{ "Name": "Bob", "Age": 42, "Height": "6 ft" }""" = Ok (Pe
...
+
+Json.serialize<Person2> (Person1 { Name = "Bob"; Age = 42 }) = Error "Incorrect record type"
+
+type StringResult<’a> = Result<’a, String>
+Json.serialize<StringResult<Int>> (Ok 1) = Ok "{\"Ok\":1}"
+Json.serialize<StringResult<Int>> (Error “err”) = Ok "{\"Error\":\"err\"}"
+Json.serialize<StringResult<Int>> (Error 1) = Error "Incorrect type for Error value"
+
+// TODO: This test should fail but doesn't
+// Json.serialize<StringResult<Int>> (Error 1) = Ok "{\"Error\":1}"
```
Comment (OceanOak) [inline backend/testfiles/execution/json.tests:344]: Should we follow-up on this in a different PR?

Comment (StachuDotNet) [reply backend/testfiles/execution/json.tests:344]: Yeah - respecting type args feels out of scope for this PR, though we do need to follow up on that before too long.

File: backend/src/StdLibExecution/Libs/Json.fs [hunk 4]
```diff
@@ -428,6 +420,7 @@ let parse
           |> Map.ofSeq
 
         DRecord(typeName, dvalMap)
+      | _ -> Exception.raiseInternal "TODO: Error message" []
```
Comment (StachuDotNet) [inline backend/src/StdLibExecution/Libs/Json.fs:423]: worth an update

File: backend/src/StdLibExecution/Libs/Json.fs [hunk 5]
```diff
@@ -122,76 +122,66 @@ let rec serialize
...
+
+        let fieldDefs = matchingCase.fields |> List.map (fun def -> def.typ)
+        // TODO: ensure that the field count is the same
+
+        w.writeObject (fun () ->
+          w.WritePropertyName caseName
+          w.writeArray (fun () ->
+            List.zip fieldDefs fields
+            |> List.iter (fun (fieldDef, fieldVal) -> r fieldDef fieldVal)))
+      | _ -> Exception.raiseInternal "TODO: error message" []
```
Comment (StachuDotNet) [inline backend/src/StdLibExecution/Libs/Json.fs:151]: and here

File: backend/src/LibBackend/SqlCompiler.fs [hunk 6]
```diff
@@ -445,6 +445,11 @@ let rec lambdaToSql
         // TYPESCLEANUP use args
         | TCustomType (typeName, args) ->
           match Map.get typeName types with
+          // TODO: Deal with alias of record type
+          | Some (CustomType.Alias _) ->
+            error2
+              "The datastore's type is not a record"
+              (FQTypeName.toString typeName)
```
Comment (StachuDotNet) [inline backend/src/LibBackend/SqlCompiler.fs:452]: Let's add a task around this in "backend work to be done" in notion

**Review by StachuDotNet — APPROVED** [review]
There might be use of additional alias tests (worth asking GPT for ideas?) but barring the minor comments, this looks great.

File: backend/src/LibBackend/SqlCompiler.fs (hunk 6 above)
Comment (OceanOak) [reply backend/src/LibBackend/SqlCompiler.fs:452]: [Done](https://www.notion.so/darklang/backend-work-to-be-done-0c191de14aed45d8892a3d289d0b5885?pvs=4#76375858b1e4496e86d752dfe8966cde)

**Comment by OceanOak** [comment]
If you think we need more tests please let me know.

**Pull request merged**